### PR TITLE
fix: respect `webpack` aliases (`resolve.alias`)

### PR DESCRIPTION
**Pull request by evilebottnawi**
Fixes https://github.com/webpack-contrib/sass-loader/issues/410

### CODE REVIEW

**Comment by michael-ciniawsky** [comment]
fix or better feature ? :)

**Comment by evilebottnawi** [comment]
@michael-ciniawsky more fix, than feature, webpack by default support aliases, need sombody who know wepback and sass to be sure that all good

**Comment by d3viant0ne** [comment]
@evilebottnawi - @jhnns has been busy but he's still available when needed. Just ping him on slack like you would Tobias if you need him to look at a PR.

File: test/sass/import-css.sass [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // Special behavior of node-sass/libsass with CSS-files
 // 1. CSS-files are not included, but linked with @import url(path/to/css)
-@import ../node_modules/css/some-css-module.css
+@import ~css/some-css-module.css
```
Comment (jhnns) [inline test/sass/import-css.sass]: Thanks for spotting this. Seems like it slipped through with some of the last commits...

**Comment by evilebottnawi** [comment]
@jhnns big bad problem with aliases is nonsupporting this inside `node_modules`, because we can't add `~` inside other libraries/frameworks. To be honest, I do not know how to solve this, in theory we can add option where developers will add own aliases, anywhere here https://github.com/webpack-contrib/sass-loader/blob/master/lib/importsToResolve.js#L16. Example:
```shell
{
                loader: "sass-loader",
                options: {
                    includePaths: ["absolute/path/a", "absolute/path/b"],
                    aliases: { /regex/: 'path/to/something' } // as in webpack
                }
            }
```
I agree this look strange, but we can explain in `README` why we use this solution. I can rewrite this here if we will accept this. Thanks!

**Comment by jhnns** [comment]
@evilebottnawi Sorry for being unresponsive.

I won't add any additional custom aliasing to the sass-loader since it is already complicated enough. If webpack's `module.alias` option would be more powerful, this could be solved by anyone.

However, I think you use case should be solvable today by writing [a custom importer](https://github.com/sass/node-sass#importer--v200---experimental) for this particular file:

```js
const nodeSass = require("node-sass");

// webpack.config.js
module.exports = {
    ...
    {
        loader: "sass-loader",
        options: {
            importer: [
                // url will be the string passed to @import
                // prev is the file where the import was encountered
                (url, prev) => (
                    shouldBeAliased(url) ?
                        { file: require.resolve("path/to/alias.sass") } :
                        // pass file to the next importer
                        // the last importer is the webpack importer inserted by the sass-loader
                        nodeSass.types.Null()
                )
            ]
        }
    }
};
```

I haven't tested this, but this *should* work. Please let me know if it actually did 😁. Or try out the [node-sass-magic-importer](https://www.npmjs.com/package/node-sass-magic-importer) which may solve this for you already...

You can also write a custom resolver for webpack, but this is more complicated:

```js
class ResolverPlugin {
    constructor(options) {
        this.options = options;
    }
    apply(resolver) {
        resolver.plugin("described-resolve", (request, callback) => {
            // request is an object that contains meta information about the resolve request
            // request.request is a string that has been passed to the resolver, similar to url in the custom importer
            // request.context.issuer is an absolute path to the file that wants to import the request

            if (shouldBeAliased(request.request) === false) {
                callback();

                return;
            }

            resolver.doResolve(
                "resolve",
                Object.assign({}, request, {
                    request: require.resolve("path/to/alias.sass")
                }),
                "aliased sass file",
                callback
            );
        });
    }
}

module.exports = class AliasSassFilePlugin {
    constructor(options) {
        this.options = options;
    }

    apply(compiler) {
        compiler.plugin("after-resolvers", () => {
            compiler.resolvers.normal.apply(new ResolverPlugin(this.options));
        });
    }
}
```

**Comment by mischkl** [comment]
@evilebottnawi @jhnns As I mentioned in #466 this means webpack is essentially incompatible with a large segment of third-party scss modules on NPM, including [material components](http://github.com/material-components/material-components-web) from Google. For anyone who wants to use such packages to need to build in this kind of hack seems like a poor solution. Especially considering that users of e.g. `@angular/cli` have no such ability to configure their build in this way.

What we really need IMHO is for sass-loader to process imports that occur within third-party modules differently than those that occur in the webpack project itself. For reference, everything that [https://github.com/sass-eyeglass/eyeglass](Eyeglass) supports should ideally be able to be resolved by sass-loader, for the purpose of long-term interoperability of the third-party scss module ecosystem.

**Comment by evilebottnawi** [comment]
@mischkl Be great if your do PR, @jhnns now is very busy and only do review PR, i am not familiar with `eyeglass` and for me difficult understand what need to add

**Comment by michael-ciniawsky** [comment]
What's the status here ? :)

**Comment by evilebottnawi** [comment]
@michael-ciniawsky need answer from @jhnns

**Pull request merged**